src/engine/structures.rs

```diff
@@ -2,62 +2,121 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-use std::collections::HashMap;
+use std::collections::{HashMap, hash_map};
 use std::iter::IntoIterator;
-use std::slice::{Iter, IterMut};
 
 use uuid::Uuid;
 
-use super::engine::{HasName, HasUuid};
-
+use engine::Name;
 
 /// Map UUID and name to T items.
 #[derive(Debug)]
-pub struct Table<T: HasName + HasUuid> {
-    items: Vec<T>,
-    name_map: HashMap<String, usize>,
-    uuid_map: HashMap<Uuid, usize>,
+pub struct Table<T> {
+    name_to_uuid: HashMap<Name, Uuid>,
+    items: HashMap<Uuid, (Name, T)>,
 }
 
-impl<T: HasName + HasUuid> Default for Table<T> {
+
+impl<T> Default for Table<T> {
     fn default() -> Table<T> {
         Table {
-            items: Vec::default(),
-            name_map: HashMap::default(),
-            uuid_map: HashMap::default(),
+            name_to_uuid: HashMap::default(),
+            items: HashMap::default(),
         }
     }
 }
 
-impl<'a, T: HasName + HasUuid> IntoIterator for &'a mut Table<T> {
-    type Item = &'a mut T;
-    type IntoIter = IterMut<'a, T>;
+pub struct Iter<'a, T: 'a> {
+    items: hash_map::Iter<'a, Uuid, (Name, T)>,
+}
 
-    fn into_iter(self) -> IterMut<'a, T> {
-        self.items.iter_mut()
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = (&'a Name, &'a Uuid, &'a T);
+
+    #[inline]
```

Comment:
Could you leave ```inline``` hints out, unless you're absolutely convinced that they are desirable? We don't do this anywhere else.

Reply:
This is what other container implementations do, so I'm gonna leave it.